src/pyhf/events.py

```diff
@@ -24,19 +23,41 @@ def noop(*args, **kwargs):
     pass
 
 
-class WeakList(list):
-    def append(self, item):
-        list.append(self, weakref.WeakMethod(item, self.remove))
```

Comment:
I need to teach myself about [`weakref.WeakMethod`](https://docs.python.org/3/library/weakref.html#weakref.WeakMethod) but I'm not clear on why we won't need to do a `self.remove` in the `Callables` append method.

Reply:
We could, but relying on the callback of something to be deleted is a bit trickier when iterating over loops where something gets deleted during loop iteration. It's generally going to be easier to flush things out every time you access something in the weak list.